File: Algorithm/C++/1000+/3356. Zero Array Transformation II.cc
#include <vector>

class Solution {
public:
    int minZeroArray(std::vector<int>& nums, std::vector<std::vector<int>>& queries) {
        int n = nums.size();
        int idx = 0, curr_sum = 0;
        std::vector<int> counter(n+1, 0);
        for(int i = 0; i < n; i++) {
            curr_sum += counter[i];
            while(curr_sum < nums[i] && idx < queries.size()) {
                int l = queries[idx][0], r = queries[idx][1], v = queries[idx][2];
                counter[l] += v;
                counter[r+1] -= v;
                if (l <= i && i <= r) {
                    curr_sum += v;
                }
                idx++;
            }
            if (idx == queries.size() && curr_sum < nums[i]) {
                return -1;
            }
        }
        return idx;
    }
};

File: Algorithm/Python/1000+/3356. Zero Array Transformation II.py
class Solution(object):
    def minZeroArray(self, nums, queries):
        """
        :type nums: List[int]
        :type queries: List[List[int]]
        :rtype: int
        """
        curr_sum, idx = 0, 0
        count = [0] * (len(nums)+1)
        for i in range(len(nums)):
            curr_sum += count[i]
            while nums[i] > curr_sum and idx < len(queries):
                l, r, v = queries[idx]
                count[l] += v
                count[r+1] -= v
                if l <= i and i <= r:
                    curr_sum += v
                idx += 1
            if idx == len(queries) and curr_sum < nums[i]:
                return -1
        return idx